File: Java/search/Fibonacci_Search.java
/* Fibonacci search works for sorted array,
   it is basically a comparison-based techhnique
   that uses fibonacci numbers to search an element in an array.
*/

import java.util.*;
//Declaring fibonacci_Search class
class Fibonacci_Search
{
    //Method to find minimum in two numbers
    public static int min(int x,int y)
    {
        return (x<=y?x:y);
    }
    //Fibonacci serach method
    public static int fibonacci_Search(int arr[],int s,int n)
    {
	    int fib2=0;
	    int fib1=1;
	    int fib=fib2+fib1;
	    while(fib<n)
	    {
		    fib2=fib1;
		    fib1=fib;
		    fib=fib2+fib1;
	    }
	    int offset=-1;
	    while(fib>1)
	    {
		    int i=min(fib2+offset,n-1);
		    if(arr[i]<s)
		    {
			    fib=fib1;
			    fib1=fib2;
			    fib2=fib-fib1;
			    offset=i;
		    }
		    else if(arr[i]>s)
		    {
			    fib=fib2;
			    fib1=fib1-fib2;
			    fib2=fib-fib1;
		    }
		    else
			    return i;
	    }
		if(fib1==s && arr[offset+1]==s)
			return offset+1;
		return -1;
}
//driver code
public static void main(String args[])
{
    int n,s;
    System.out.println("Fibonacci Search program:");
    System.out.println("Enter the size of an array:");
    Scanner sc=new Scanner(System.in); 
    n=sc.nextInt();  
    int[] arr = new int[n];
    System.out.println("Enter the elements of the array:");  
    for(int i=0; i<n; i++)  
    {  
        //reading array elements from the user   
        arr[i]=sc.nextInt();  
    }   
    System.out.println("Enter the elements to search in array:");
    s=sc.nextInt();
    //calling function and printing output
    int index=fibonacci_Search(arr, s, n);
    if(index>=0)
        System.out.print("Found at index:" +index);
    else
        System.out.print(s+" isn't present in the array.");
}
}

/* Input_1:
Enter the size of an array:
5
Enter the elements of an array:
25
35
45
65
85
Enter the elements to search in array:
65
Output_1:
Found at index:3

Input_2:
Enter the size of an array:
5
Enter the elements of an array:
25
35
45
65
85
Enter the elements to search in array:
55
Output_2:
55 isn't present in the array.
*/

File: Julia/search/fibonacci_search.jl
"""Julia program to implement Fibonacci Search algorithm.
Fibonacci Search is a Divide and Conquer strategy based algorithm that works for sorted arrays
Fibonacci Search avoids CPU costly division operation and uses addition and subtraction opertations
This algorithm uses Fibonacci series as an underlying principle, where every element of the series is defined by:

    fib(n) = fib(n-1) + fib(n-2), where fib(0) = 0, fib(1) = 1
"""

function fibonacci_search(arr, n, ele)

    # We start the series from 1, 1 as the Julia language is 1-based indexing
    fibn2 = 1
    fibn1 = 1
    fibn = fibn1 + fibn2
    while (fibn < n)
        fibn2 = fibn1
        fibn1 = fibn
        fibn = fibn1 + fibn2
    end

    # To denote the eliminted range from left.
    offset = 0

    while (fibn > 1)
        i = min(offset + fibn2, n - 1)
        
        """Compares the element in i'th index with ele and decide the next fibonacci element. If the element doesn't
            turns out to be greater than or less than the element at i'th index then we find the required element"""
        if (arr[i] < ele) 
            fibn = fibn1;
            fibn1 = fibn2;
            fibn2 = fibn - fibn1;
            offset = i;
        elseif (arr[i] > ele) 
            fibn = fibn2;
            fibn1 = fibn1 - fibn2;
            fibn2 = fibn - fibn1;
        else
            return true
        end
        if (arr[offset + 1] == ele)
            return true
        end
    end
    return false
end


print("How many numbers are present in the array? ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("Array is Empty!!!")
    exit()
end
arr = Int[]
print("Enter the numbers: ")
arr = [parse(Int, num) for num in split(readline())] 
print("Which number do you want to search in the array? ")
ele = readline()
ele = parse(Int, ele)
# Sort the array in ascending order
arr = sort(arr)
res = fibonacci_search(arr, n, ele)
if (res == 0)
    print("The number $ele is not present in the array")
else
    print("The number $ele is present in the array.")
end


"""
Time Complexity - O(log(n)), where 'n' is the size of the array
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I

How many numbers are present in the array? 5 
Enter the numbers: 1 2 3 4 5
Which number do you want to search in the array? 6
The number 6 is not present in the array

SAMPLE II

How many numbers are present in the array? 3
Enter the numbers: 3 1 2
Which number do you want to search in the array? 2
The number 2 is present in the array.

"""
